### ISSUE TITLE
[v2.2 bug] 'this' lost in default functions of props

### ISSUE THREAD

**Issue by beeplin** [vuejs/vue#5077]
The following code seems broken in v2.2.1:

```js
  props: {
    goBack: {
      type: Function,
      'default': function() {
        this.$router.go(-1);
      }
    }
  }
```
error message: 
```
Uncaught TypeError: Cannot read property '$router' of null
    at default ......
```

**Comment by Kingwl** [comment]
https://jsfiddle.net/z11fe07p/858/ 
It looks like it works fine

**Comment by beeplin** [comment]
Yeah, I just tested in jsfiddle and it's fine.  My fault. Sorry to bother.

BTW this is why I got this error: if we have a method named 'methodFunc' in a vue instance, we can call it in template either with `<button @click="methodFunc()">` or with  `<button @click="methodFunc">` . It's ok to omit the `()`. But if we have a prop named 'propFunc' whose type is function, we can only call it with `<button @click="propFunc()">`, and the `()` cannot be omitted.
